python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -222,54 +225,62 @@ def _update(self):
             nodes_allowed_to_terminate = self._get_nodes_allowed_to_terminate(
                 sorted_node_ids)
 
+        nodes_to_keep = []
+
+        def keep_node(node_id) -> None:
+            # Update per-type counts.
+            tags = self.provider.node_tags(node_id)
+            if TAG_RAY_USER_NODE_TYPE in tags:
+                node_type = tags[TAG_RAY_USER_NODE_TYPE]
+                node_type_counts[node_type] += 1
+            nodes_to_keep.append(node_id)
+
+        def terminate_node(node_id, kind, why=None) -> None:
+            why = why or kind
+            logger.info("StandardAutoscaler: "
+                        "{}: Terminating {} node.".format(node_id, kind))
+            self.event_summarizer.add(
+                "Removing {} nodes of type " + self._get_node_type(node_id) +
+                " ({}).".format(why),
+                quantity=1,
+                aggregate=operator.add)
+            nodes_to_terminate.append(node_id)
+
+        nodes_we_could_terminate = []
+
         for node_id in sorted_node_ids:
             # Make sure to not kill idle node types if the number of workers
             # of that type is lower/equal to the min_workers of that type
             # or it is needed for request_resources().
-            should_keep_worker_of_node_type, node_type_counts = \
-                self._keep_worker_of_node_type(node_id, node_type_counts)
-            if ((should_keep_worker_of_node_type
+            should_keep_or_terminate = self._keep_worker_of_node_type(
+                node_id, node_type_counts)
+            if ((should_keep_or_terminate == KeepOrTerminate.keep
                  or not nodes_allowed_to_terminate.get(node_id, True))
                     and self.launch_config_ok(node_id)):
+                keep_node(node_id)
+                continue
+            if should_keep_or_terminate == KeepOrTerminate.terminate:
+                terminate_node(node_id, "max_workers_per_type")
                 continue
 
             node_ip = self.provider.internal_ip(node_id)
             if node_ip in last_used and last_used[node_ip] < horizon:
-                logger.info("StandardAutoscaler: "
-                            "{}: Terminating idle node.".format(node_id))
-                self.event_summarizer.add(
-                    "Removing {} nodes of type " + self._get_node_type(node_id)
-                    + " (idle).",
-                    quantity=1,
-                    aggregate=operator.add)
-                nodes_to_terminate.append(node_id)
+                terminate_node(node_id, "idle")
```

Comment:
reading terminate_node confused me a bit because this is not terminating, this is enqueueing the list.
We try to terminate in batches to reduce the overhead of terminating one by one.

Reply:
Maybe considering renaming?

Reply:
good point